temba/channels/tests.py

```diff
@@ -8187,3 +8212,74 @@ def test_send(self):
                 self.assertEquals(ERRORED, msg.status)
                 self.assertEquals(1, msg.error_count)
                 self.assertTrue(msg.next_attempt)
+
+
+class FcmTest(TembaTest):
+
+    def setUp(self):
+        super(FcmTest, self).setUp()
+
+        self.channel.delete()
+        self.channel = Channel.create(self.org, self.user, None, Channel.TYPE_FCM, 'FCM Channel', 'fcm-channel',
+                                      config=dict(FCM_KEY='123456789', FCM_TITLE='FCM Channel'),
+                                      uuid='00000000-0000-0000-0000-000000001234')
+
+    def test_receive(self):
+        data = {'from': '12345abcde', 'msg': 'Hello World!'}
+        callback_url = reverse('handlers.fcm_handler', args=['receive', self.channel.uuid])
+        response = self.client.post(callback_url, data)
+        self.assertEquals(400, response.status_code)
+
+        data = {'from': '12345abcde', 'msg': 'Hello World!', 'fcm_token': '1234567890qwertyuiop'}
+        callback_url = reverse('handlers.fcm_handler', args=['receive', self.channel.uuid])
+        response = self.client.post(callback_url, data)
+        self.assertEquals(200, response.status_code)
+
+        # load our message
+        msg = Msg.objects.get()
+        self.assertEquals("1234567890qwertyuiop", msg.contact.get_urn(FCM_SCHEME).path)
+        self.assertEquals("fcm:12345abcde", msg.contact.get_urn(FCM_SCHEME).urn)
+        self.assertEquals(self.org, msg.org)
+        self.assertEquals(self.channel, msg.channel)
+        self.assertEquals("Hello World!", msg.text)
+
+    def test_register(self):
+        data = {'urn': '12345abcde'}
+        callback_url = reverse('handlers.fcm_handler', args=['register', self.channel.uuid])
+        response = self.client.post(callback_url, data)
+        self.assertEquals(400, response.status_code)
+
+        data = {'urn': '12345abcde', 'fcm_token': '1234567890qwertyuiop'}
+        callback_url = reverse('handlers.fcm_handler', args=['register', self.channel.uuid])
+        response = self.client.post(callback_url, data)
+        self.assertEquals(200, response.status_code)
+
+    def test_send(self):
+        joe = self.create_contact("Joe", urn="fcm:12345abcde", extra_path="123456abcdef")
+        msg = joe.send("Hello, world!", self.admin, trigger_send=False)
+
+        with self.settings(SEND_MESSAGES=True):
+
+            with patch('requests.post') as mock:
+                mock.return_value = MockResponse(200, '{ "success": 1, "multicast_id": 123456, "error": 0 }')
+
+                # manually send it off
+                Channel.send_message(dict_to_struct('MsgStruct', msg.as_task_json()))
+
+                # check the status of the message is now sent
+                msg.refresh_from_db()
+                self.assertEqual(msg.status, WIRED)
+                self.assertTrue(msg.sent_on)
+                self.clear_cache()
```

Comment:
Good to check that the post was made with the arguments you expect. (see `mock.assert_called_with`)